### ISSUE TITLE
Introduce new active consumers entry point class and storage infrastructure

### ISSUE THREAD

**Issue by felixarntz** [google/site-kit-wp#5567]
In order to keep track of active consumers of an access token (as part of dashboard sharing), the necessary infrastructure needs to be created. See [design doc](https://docs.google.com/document/d/1KODBTcjRjDNNgYd4Qv-d0XlapdYkmBdyzqNTFmEg9Rg/edit).

This is a priority for dashboard sharing, but not a launch blocker.

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance criteria

* A new PHP class for an `active_consumers` user meta key should be introduced, extending `User_Setting`.
    * The meta value should allow for an array of associative arrays to be stored, each with `id` (integer) and `roles` (array of strings) keys.
    * The default value should be the empty array.
    * See [design doc section for details](https://docs.google.com/document/d/1KODBTcjRjDNNgYd4Qv-d0XlapdYkmBdyzqNTFmEg9Rg/edit?resourcekey=0-eZkvc2ChBvVP_U_p7DdtwQ#heading=h.qjeawqqo05cs).
* A new PHP entry point class for managing active consumers and all related functionality should be introduced, including a `register` method as usual.
    * For now, all it should do is register the user meta key via the above class.
    * The class should be instantiated and initialized in `Plugin`, similar to other entry points.

## Implementation Brief

- In `includes/Core` folder, create a new folder named `Dashboard_Sharing` and inside it, create another folder named `Activity_Metrics`. Inside this folder:
    - Create a new file named `Active_Consumers.php`. In this file:
        - Create a class named `Active_Consumers` extending the `Google\Site_Kit\Core\Storage\User_Setting` class.
        - Inside the class:
            - Create a new constant named `OPTION` with the value `googlesitekit_active_consumers`.
            - Create a protected function named `get_type`, which should return a string containing `array`. This sets the type of this user meta to an array.
            - Create a protected function named `get_default`, which should return an empty array. This sets the default value of this user meta to an empty array.
        - Set the namespace, file headers, and docblock comments appropriately. Examples can be taken from `includes/Core/Feature_Tours/Dismissed_Tours.php`.
    - Create a new file named `Activity_Metrics.php`. In this file:
        - Create a class named `Activity_Metrics`.
        - Inside the class:
            - Create a protected property named `$active_consumers` which is responsible for storing the `Core\Dashboard_Sharing\Activity_Metrics\Active_Consumers` class instance.
            - Create the class constructor method. In it:
                - Pass instances of `Google\Site_Kit\Context` as `$context` and `Google\Site_Kit\Core\Storage\User_Options` as `$user_options` as parameters.
                - Assign a new instance of the `Core\Dashboard_Sharing\Activity_Metrics\Active_Consumers` class to the `$active_consumers` property. The class initialisation should receive `$user_options` or it not set, a new instance of `Core\Storage\User_Options` with `$context` passed to it.
            - Create a public method named `register`, which should call the `register` method of the `$active_consumers` property.
        - Set the namespace, file headers, and docblock comments appropriately. Examples can be taken from `includes/Core/Feature_Tours/Feature_Tours.php`.
- In `includes/Plugin.php`:
    - Locate the callback function for the `init` hook (i.e. `add_action( 'init' )`).
    https://github.com/google/site-kit-wp/blob/2cbb5a4c41caa1bb78f56bf30df302620bbb5adf/includes/Plugin.php#L152-L222
    - Inside this function, initialise the `Core\Dashboard_Sharing\Activity_Metrics\Activity_Metrics` class which should receive the `$context` property and `Core\Storage\User_Options` class instance (accessible via `$user_options` in this function) as the constructor argument and call the `register` method of this initialised class.

### Test Coverage

- No tests need to be added/updated at this stage.

## POC PR
[This PoC PR](https://github.com/google/site-kit-wp/pull/5663) can be used as a starting point for this issue. It covers most of the implementation mentioned above except:
- It needs to be thoroughly tested.
- The file headers, DocBlocks and other comments need to be double-checked/added.
- It needs to be ensured that there are no errors/warnings.

### QA:Eng
- If the user meta is successfully registered, it will be added to the `$wp_meta_keys` global array. To verify:
    - Use `var_dump` within the plugin to see the contents of the `$wp_meta_keys` global array (use the `global` keyword to access it).
    - In the returned array, look for `{$table_prefix}_googlesitekit_active_consumers` and ensure that it exists.

## Changelog entry

* N/A

**Comment by techanvil** [comment]
Hi @nfmohit, this is a good start, but a couple of points need to be addressed in the IB:

It looks to me like the specced `add` method is going beyond the scope of the AC. The logic that has been proposed for `add` should in fact be specced and implemented in the subsequent issue, #5568. Also, note how #5568 states:

> - Particular attention should be paid to the queries to use for this:
>   - It needs to be avoided that the implementation results in a database query running every time that the shared dashboard is accessed (e.g. avoid it by using a transient).

... this implies some as yet to be determined logic for efficiently looking up the values which would quite plausibly sit outside such an `add` method. As such I would simply remove this `add` method from the IB.

Additionally the AC asks for _two_ new classes - one that extends `User_Setting` as you have specced out, but a second which is the entry point class that will contain the functionality for managing these settings, and this entry point class is what should be instantiated and registered in `Plugin.php` . This makes sense as the logic for managing transients, updating multiple user options etc will likely need to sit in this entry point class.

A simple example of this sort of entry point class is `includes/Core/Feature_Tours/Feature_Tours.php` which itself creates and registers an instance of a `User_Setting` derived class.

So, to sum it up, please could you remove the `add` method, and add a new entry point class following the example of `Feature_Tours`.

**Comment by nfmohit** [comment]
Thank you for the kind and valid review @techanvil. Updated  the IB and PoC PR.

**Comment by techanvil** [comment]
Thanks @nfmohit!

IB :white_check_mark:
